### ISSUE TITLE
Inconsistent use of array ordering

### ISSUE THREAD

**Issue by AlexisMignon** [menpo/cyvlfeat#17]
Hello, 

First thanks for the work that has been done. I had once tried to start it but gave up in front of the amount of work. 

I guess most of the wrappers have been created by translating the matlab mex files into cython. While this makes sense, the translation has to be made carefully since matlab uses array in fortran ordering. I've noticed some errors at least in two modules: sift and fisher regarding this issue.
- In sift: there should be no translation of the descriptor nor of the positions (x, y).
- In fisher: the ordering of the vectors is inverted. I think the tests are passed only because the input data were generated the same way and the problems of reshaping cancel out.

Since correcting the problems will lead to some change in the behaviour in the package I think it is better to discuss before starting sending pull requests.

**Comment by patricksnape** [comment]
Yes, I wrote the original wrapper around SIFT in a manner that was useful for my other projects. Fisher etc were contributed by other users and I must admit I am not very familiar with those pieces of code - I mostly check them for style and was happy with tests passing. 

Can you elaborate what you mean about SIFT?

**Comment by AlexisMignon** [comment]
Regarding sift the transpose operation on the descriptor , found in the matlab code,  is needed because the descriptor is actually computed on the transposed image since the matlab uses fortran ordering. However, the C code expects the usual C ordering.

This means that in the sift wrapper  in Cython: 

1. The function transposing the descriptor should not be called
2. The inversion of x, y coordinates should not be performed
3. The transformation on the sift angle should not be performed.


More generally, the fortran ordering explains why in the matlab tool box data arrays are usually given as [n_dimensions, n_samples] instead of [n_samples, n_dimensions] which is what the C code expects.

**Comment by patricksnape** [comment]
It would be great to get line links to the code so we all know what we are talking about. It does appear that the SIFT vector is transposed when it doesn't need to be - I took more care in the dense sift apparently and recognised that that didn't need to be [done](https://github.com/menpo/cyvlfeat/blob/master/cyvlfeat/sift/cysift.pyx#L125).

With regards to the inversion of `(x, y)`, I assume you mean [this line](https://github.com/menpo/cyvlfeat/blob/master/cyvlfeat/sift/cysift.pyx#L331)? That is actually intentional as I've taken to maintaining the contract of 'first axis' and 'last axis' as opposed to `x` and `y`. This means that indexing makes far more sense because you index into the array rather than the 'image' -> `image[y, x] === image[coord[0], coord[1]]` rather than always have to flip the coordinates.

**Comment by AlexisMignon** [comment]
In the non dense version, the transposition of the descriptor is done [here](https://github.com/menpo/cyvlfeat/blob/master/cyvlfeat/sift/cysift.pyx#L311), and the angle inversion is done [here](https://github.com/menpo/cyvlfeat/blob/master/cyvlfeat/sift/cysift.pyx#L334).

As for the x, y convention, I see it's indicated in the documentation so it's faire enough I guess.

**Comment by patricksnape** [comment]
Does the 'flipping' of the angles really involve fortran ordering? Doesn't that just have to do with which way you think is 'up'?

**Comment by AlexisMignon** [comment]
The flipping comes from the fact the angles are measured on the transposed image. As a reference you can have a look at the code in the [command line program](https://github.com/vlfeat/vlfeat/blob/master/src/sift.c#L741). There is no flipping.

**Comment by patricksnape** [comment]
I agree - clearly this was an oversight on my part and it does look as though the current SIFT features are incorrect. We could even support a flag that allows one to choose whether they want x-axis or y-axis first so that the exact output can be maintained?

**Comment by patricksnape** [comment]
PR is up for SIFT - I'm unsure about Fisher - there doesn't appear to be any reshaping inside it?

**Comment by AlexisMignon** [comment]
I think the option for the x, y ordering is a nice idea. I have started to modify some code on my side but ill have to see how to reorganise my branches to be able to make separated pull requests.

**Comment by patricksnape** [comment]
Right - well then we need to ping @williamkreamer given that he contributed this and that change would break his code. @williamkreamer Would you be fine with this change which would simply require you to either transpose your input matrix or just build the matrix transposed in the first place?

**Comment by AlexisMignon** [comment]
An alternative solution can be two keep both version (I'm about sure the original version will not pass the test case though) call the new version with _transposed suffix and add a deprecation warning in original version saying it will be replaced by the transposed in a future version ?

**Comment by williamkreamer** [comment]
Guys I will have time to take a look at this over the weekend. Can you wait
until Monday?

On Fri, Jun 3, 2016 at 8:48 AM Patrick Snape <notifications@github.com>
wrote:

> Right - well then we need to ping @williamkreamer
> <https://github.com/williamkreamer> given that he contributed this and
> that change would break his code. @williamkreamer
> <https://github.com/williamkreamer> Would you be fine with this change
> which would simply require you to either transpose your input matrix or
> just build the matrix transposed in the first place?
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/menpo/cyvlfeat/issues/17#issuecomment-223569647>, or mute
> the thread
> <https://github.com/notifications/unsubscribe/ABbH7nR37i5H713FuDRjlKJ056TptFldks5qICKMgaJpZM4IsS_w>
> .
>
-- 
William Kreamer

Senior Data Scientist at TrueMotion

www.goTrueMotion.com <http://www.gotruemotion.com/>

Cell | 617.335.3858

Twitter | @goTrueMotion <http://www.twitter.com/gotruemotion>

LinkedIn | /in/wkreamer <https://www.linkedin.com/in/wkreamer>

**Comment by patricksnape** [comment]
Monday is fine with me - just don't want to pull the rug out from under you!

**Comment by williamkreamer** [comment]
I took a took a look at the fisher code. I don't see any error regarding transposition of dimensions. The implementation matches the documentation provided with the MATLAB wrapper (which itself is just a wrapper around a C function, vl_fisher_encode, see http://www.vlfeat.org/api/fisher_8c.html#a4c13fe11e9847f9046c2636a6e77c1bd)

Moreover the unit test (https://github.com/menpo/cyvlfeat/blob/master/cyvlfeat/fisher/tests/fisher_test.py#L42) has a check against results obtained by directly calling the native C vl_fisher_encode function).

So in summary, I don't see that there is an issue. Maybe there is, and I don't see it. Can either of you explain it to me?

**Comment by patricksnape** [comment]
I think @AlexisMignon is just saying that it is more common in C ordering to have your matrices in `(N, D)` ordering where `N` is the number of samples and `D` is the dimensionality. Whereas in Matlab, similar to how you would write it in a paper, the more common ordering is `(D, N)`.  So I think that @AlexisMignon is just suggesting that we maintain the more common ordering (which is used in projects such as [scikit-learn](http://scikit-learn.org/stable/modules/generated/sklearn.linear_model.LinearRegression.html#sklearn.linear_model.LinearRegression.fit)).

**Comment by williamkreamer** [comment]
BTW a point of information. If you were to provide the matrix as (n_samples, n_features) (i.e. transposed) then you would have to transpose the matrix again before calling the C function vl_fisher_encode. That C function expects the data ordered as (n_features, n_samples)

**Comment by patricksnape** [comment]
The C function just expects a linear array and then you to tell it how many dimensions/clusters/samples there are. So, as long as you pass the correct values into those no sort of transpose is required.

**Comment by patricksnape** [comment]
You [are totally correct](http://www.vlfeat.org/api/fisher_8c.html#a4c13fe11e9847f9046c2636a6e77c1bd). @AlexisMignon it looks like this change would be pointless given that the C code expects the data in `(D, N)` ordering. To quote the linked doc:

> data has dimension rows and numData columns

**Comment by AlexisMignon** [comment]
The C code does not expect the data in (D, N) ordering. It expects the data as
``` [x1, y1, z1, x2, y2, z2,]```. 

In matlab, since it uses the Fortran array ordering the above array corresponds to a ```(D, N)``` containging:

```
x1 x2
y1 y2
z1 z2
```
while it corresponds in C to a ```(N, D)``` array of containing:
```
x1 y1 z1
x2 y2 z2
```

This is why the matlab array is passed as (D, N) and the C array should be passed as (N, D).

You can have a look at the C code in vlfeat for instance:

https://github.com/vlfeat/vlfeat/blob/master/vl/fisher.c#L443

The difference between the data point and the cluster mean for the component `dim` is computed as:

```TYPE diff = data[i_d*dimension + dim] - means[i_cl*dimension + dim] ;
```
you can see that the ordering in the 1d array is expected is expected as I mentioned earlier.

So I guess that if you passed the same array the same way to the cython code and the C code you obtained the same result but it is not the expected one.

**Comment by williamkreamer** [comment]
@AlexisMignon let me take another look at the documentation and the code you cited. I'll get back to you

**Comment by williamkreamer** [comment]
@patricksnape @AlexisMignon we reviewed the vl_fisher_encode function and we agree that the data should be transposed. Can I take on a PR for this? I'll do that change today.

Specifically, I would:
1) change the docs to indicate:
 - the shape of the means, covariances is [K,D] where K is the number of GMM components and D is the dimensionality of each vector
 - the shape of the priors is [K]
 - the shape of the data matrix x is  [N,D] where N is the number of data vectors and D is the dimensionality of each vector
2) update the data shape checks appropriately in the python and the cython code
3) update the unit test comments to transpose the shapes (but won't change any of the numbers)

**Comment by williamkreamer** [comment]
wait :) I see you already have a PR where you've done all of this and more - https://github.com/menpo/cyvlfeat/pull/20

Cool. So will you merge this into master soon? Thanks for catching the error @AlexisMignon

**Comment by williamkreamer** [comment]
BTW - and maybe this isn't important - it looks to me like the vlfeat documentation is wrong (wrt the note about column-major ordering).

We've agreed that the shape of the input data is [N rows by D columns] where N is the number of vectors and D is the number of dimensions in each vector. And that the [C code](https://github.com/vlfeat/vlfeat/blob/master/vl/fisher.c#L443) reads out contiguous elements from the input matrix for each dimension of the input vectors.

Isn't that the definition of [row-major](https://en.wikipedia.org/wiki/Row-major_order) ordering? In which case the vlfeat documentation is wrong (the note about column-major ordering)?

**Comment by patricksnape** [comment]
Actually that code is agnostic to the ordering due to the `i_d * dimension + dim`. If you change the value of `dimension` and `dim` you can 'simulate' either row-major or column-major. It just matters how many elements you 'skip' and how many elements you think are contiguous.

**Comment by williamkreamer** [comment]
Huh. I'll have to think about that. So was this change necessary in the end
because of an error? Or was it just a choice of conventions?
On Tue, Jun 7, 2016 at 4:10 AM Patrick Snape <notifications@github.com>
wrote:

> Actually that code is agnostic to the ordering due to the i_d * dimension
> + dim. If you change the value of dimension and dim you can 'simulate'
> either row-major or column-major. It just matters how many elements you
> 'skip' and how many elements you think are contiguous.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/menpo/cyvlfeat/issues/17#issuecomment-224210118>, or mute
> the thread
> <https://github.com/notifications/unsubscribe/ABbH7m-qxpbZOsH_NrCvkjpWrE2QQIMKks5qJSd7gaJpZM4IsS_w>
> .
>
-- 
William Kreamer

Senior Data Scientist at TrueMotion

www.goTrueMotion.com <http://www.gotruemotion.com/>

Cell | 617.335.3858

Twitter | @goTrueMotion <http://www.twitter.com/gotruemotion>

LinkedIn | /in/wkreamer <https://www.linkedin.com/in/wkreamer>

**Comment by patricksnape** [comment]
It was just a choice of conventions - your code was still correct. It was merely that @AlexisMignon wanted to make the notation consistent across all the modules and other functions such as dense sift already expected the C-ordering. To clarify about the stride pattern:

```
D = 3  # Dimensions
N = 2  # Samples (let's call them two 3D points)
X = [a b c]
    [d e f]
X.shape == (N, D)
```

Row Major (C)
--------------------
```
vec(X) = [a b c d e f]
```
To linearly index the samples (with an index `i`), we want to recover `a b c` and then `d e f`. Therefore, `i` will have possible values `{0, 1} = range(N)` and `j` is the dimension index with possible values `{0, 1, 2} = range(D)` and thus linear indexing would be performed

```
for i in range(N):
  for j in range(D):
     sample_index = (i * D) + j  # X[i, j]

Out: [0, 1, 2, 3, 4, 5]
```
Which is a nice linear indexing that traverses the array in dimension length blocks.

Column Major (Fortran)
--------------------------------
```
vec(X) = [a d b e c f]
```

To linearly index the samples (with an index `i`), we want to recover `a b c` and then `d e f`. Therefore, `i` will have possible values `{0, 1} = range(N)` and `j` is the dimension index with possible values `{0, 1, 2} = range(D)` and thus linear indexing would be performed

```
for i in range(N):
  for j in range(D):
     sample_index = i + (j * N)  # X[j, i]

Out: [0, 2, 4, 1, 3, 5]
```
Which is obviously a horrible axis pattern that keeps skipping dimension length through the array. However, if we transpose `X`:

```
Y = X^T = [a c e]
          [b d f]
Y.shape == (D, N)
vec(Y) = [a b c d e f]
```

then the correct linear index is:
```
for i in range(N):
  for j in range(D):
     sample_index = (i * D) + j  # Y[j, i]

Out: [0, 1, 2, 3, 4, 5]
```
which is again nice and contiguous.

**Comment by AlexisMignon** [comment]
I don't think it was just a matter of convention. The array was given as (D x N) in C-order:

```
a = [[x1 x2 x3]
       [y1 y2 y3]] 
```
which means that in memory it was:
```
 [x1 x2 x3 y1 y2 y3]
```

while the C code expects it as:
```
a = [x1 y1 x2 y2 x3 y3]
```

**Comment by patricksnape** [comment]
Not if you swap the stride - change the size of the dimensions and it will skip through the memory differently.

**Comment by AlexisMignon** [comment]
I'm not sure about what you mean by swapping the stride. The way the C code in vlfeat strides on the memory is fixed. Or do you mean asking numpy to impose the a F-ordering ?

**Comment by patricksnape** [comment]
I mean this code: `i_d * dimension + dim`

Which determines the stride. Change the value of dimension and dim and you can 'simulate' either Fortran or C-ordering, though one will be far much efficient depending on the memory layout.

**Comment by patricksnape** [comment]
Closed by #20
